### PR TITLE
Bleutooth: Mesh: add experimental support of mbedtls psa as crypto backend

### LINKED ISSUE #23292: Bluetooth: Mesh and SMP DHKey coexistence
**Describe the bug**
Bluetooth SMP protocol and Bluetooth Mesh provisioner both use the bt_pub_key_gen API.
This API uses the HCI interface to generate a single public/private key pair. The public key is given to the remote, and the private key is used when received the remote public key to generate the DHKey.

While SMP only generates a single public key during init Mesh will generate a new one for each provisioning.

These two use-cases of the DHKey functionality in HCI cannot coexist because the HCI interface only support a single-instance public/private key-pair.
So we cannot support generating DHKeys using two different public keys

**To Reproduce**
Enable both Bluetooth Mesh and SMP protocols.
Perform provisioning.
Pairing cannot be performed anymore since private key used in DHKey is not valid for the public key SMP sent to the remote.

**Expected behavior**
If we are to claim both protocols working simultaneously the DHkey calculations have to be independent.

**Impact**
Bluetooth Mesh and SMP cannot both be enabled.

**Suggested fix**
If we wish to support Mesh simultaneous with regular Bluetooth connections we need to solve the  structural dependency of the HCI commands. This might mean we cannot  rely on the HCI interface and instead move this functionality to the host and change the existing internal ECC API.

**Environment:**
 - Commit: 1.14 or 2.1.0 (or any other commit supporting both protocols).

### PR DESCRIPTION
**Pull request by alxelax**
1. Split ble mesh crypto on security library dependent and independent parts
2. Refactor provisioning and crypto to remove dependency on host in aspect of DH key pair generation. There is no necessity to wait for HCI interface anymore, so keys and shared secret are generated instantly. This change also fixes the issue: https://github.com/zephyrproject-rtos/zephyr/issues/23292
3. Add experimental support mbedtls psa as crypto backend (run only ble mesh bsim tests).

PSA support requires additional rework key usage within the mesh stack to get rid of raw key representation and use key ID instead in case of PSA (work in progress with this).

### CODE REVIEW

**Comment by Laczen** [comment]
@alxelax, out of curiosity, what is the difference in ram/rom usage for mbedtls-psa vs tinycrypt?

**Comment by Laczen** [comment]
@alxelax, thanks for the info.

File: tests/bsim/bluetooth/mesh/overlay_low_lat.conf [hunk 1]
```diff
@@ -2,3 +2,4 @@ CONFIG_BT_EXT_ADV=n
 CONFIG_BT_CTLR_ADVANCED_FEATURES=y
 CONFIG_BT_CTLR_LOW_LAT=y
 CONFIG_BT_TICKER_LOW_LAT=y
+CONFIG_BT_MESH_FRIEND_ADV_LATENCY=2
```
Comment (PavelVPV) [inline tests/bsim/bluetooth/mesh/overlay_low_lat.conf:5]: Could you elaborate why this is needed?

File: tests/bsim/bluetooth/mesh/src/dfu_blob_common.c [hunk 2]
```diff
@@ -52,7 +52,7 @@ void common_sar_conf(uint16_t addr)
 	 */
 	struct bt_mesh_sar_tx tx_set = {
 		.seg_int_step = 1,
-		.unicast_retrans_count = 2,
+		.unicast_retrans_count = 3,
```
[the remark this answers is not in the document]
Comment (alxelax) [reply tests/bsim/bluetooth/mesh/src/dfu_blob_common.c:55]: `2` works with tinycrypt but doesn't with psa. At some stage, one segment cannot be delivered. I guess the timeline has been changed and there are more collisions. One more retransmit makes dfu test suite passable for both crypto libraries.

File: tests/bsim/bluetooth/mesh/overlay_low_lat.conf (hunk 1 above)
Comment (alxelax) [reply tests/bsim/bluetooth/mesh/overlay_low_lat.conf:5]: I elaborated it here: https://github.com/zephyrproject-rtos/zephyr/pull/56423/files#diff-c40d529b65a93a6bf2872ca227ccec6ff4a2bf8eaa0b473f8ee397dd476e3955R544-R547

`mesh_msg` test that is run in almost all configurations and this is quite fragile to timing changes. After provisioning refactoring, the provisioning itself started working faster and test became broken even for tinycrypt on the legacy advertiser with a low latency feature. Adding more latency (so seems fair enough for a low latency option) it started passing again. In general, this test should be redesigned (what is in progress right now).

File: tests/bsim/bluetooth/mesh/src/dfu_blob_common.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (alxelax) [reply tests/bsim/bluetooth/mesh/src/dfu_blob_common.c:55]: I can add conditional compilation here if you prefer library differences but I guess test might be broken after timing changes for tinycrypt too.

[the remark this answers is not in the document]
Comment (PavelVPV) [reply tests/bsim/bluetooth/mesh/src/dfu_blob_common.c:55]: How can I compile bsim tests with mbedtls on this branch?

[the remark this answers is not in the document]
Comment (alxelax) [reply tests/bsim/bluetooth/mesh/src/dfu_blob_common.c:55]: The simplest way is just this change:
```
choice BT_MESH_CRYPTO_LIB
	prompt "Crypto library selection for mesh security"
	default BT_MESH_USES_MBEDTLS_PSA
```
in KConfig(lines 16-18) in mesh

**Review by PavelVPV — APPROVED** [review]

**Review by Andrewpini — COMMENTED** [review]
I've got a couple of questions regarding some documentation. Other than that this looks good to me 👍

File: subsys/bluetooth/mesh/Kconfig [hunk 3]
```diff
@@ -17,6 +13,48 @@ menuconfig BT_MESH
...
+	select MBEDTLS_CIPHER_CCM_ENABLED
+	select MBEDTLS_ECP_C
+	select MBEDTLS_ECDH_C
+	select MBEDTLS_ECDSA_C
+	select MBEDTLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED
+	select MBEDTLS_ECP_DP_SECP256R1_ENABLED
+	select MBEDTLS_PK_WRITE_C
+	help
+	  Use mbed TLS library to perform crypto operations. Support of
+	  mbed TLS and PSA is experimental and only BabbleSim tests were run.
```
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/Kconfig:51]: Yep. The only bsim tests are possible to run with these changes.

**Review by Andrewpini — APPROVED** [review]
I'll approve. Fix the docs explanation if you want to.

**Pull request merged**